File: 백준/Silver/2579. 계단 오르기/계단 오르기.py
import sys
input = sys.stdin.readline
n = int(input())
step = [0]*301
for i in range(n):
    step[i+1] = int(input())
score = [0]*301
score[1] = step[1]
score[2] = step[1]+step[2]
score[3] = max(step[3]+step[1], step[3]+step[2])
for i in range(4, n+1):
    score[i] = max(score[i-2], score[i-3]+step[i-1])+step[i]
print(score[n])

File: 백준/Silver/2579. 계단 오르기/계단 오르기.java
import java.io.*;
import java.util.*;

public class Main {
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    static BufferedReader br =   new BufferedReader(new InputStreamReader(System.in));
    static int N;
    static int[][] d;
    static int[] steps;

    public static void main(String[] args) throws IOException {
        N = Integer.parseInt(br.readLine());
        steps = new int[N];
        for (int i = 0; i < N; i++) {
            steps[i] = Integer.parseInt(br.readLine());
        }
        d = new int[N][2];
        bw.write(String.valueOf(Math.max(f(0, 0), f(1, 0))));

        br.close();

        bw.flush();
        bw.close();
    }

    private static int f(int step, int count) {
        if (step >= N) return -100_000_000;
        if (step == N - 1) return steps[step];
        if (d[step][count] != 0) return d[step][count];
        if (count == 0) d[step][count] = Math.max(f(step + 1, 1), f(step + 2, 0)) + steps[step];
        else d[step][count] = f(step + 2, 0) + steps[step];
        return d[step][count];
    }
}
